src/evict/evict_lru.c

```diff
@@ -708,11 +709,25 @@ __evict_update_work(WT_SESSION_IMPL *session)
                 (void)ret; /* Keep the assignment to 0 just in case, but suppress clang warnings. */
                 break;
             }
-            if (__wt_hs_get_btree(session, hs_id, &hs_tree) == 0) {
+            /*
+             * At this point, we are under the evict pass lock and should only attempt to read from
+             * the cursors dhandle cache to obtain the HS. If it is not present in the cursors
+             * dhandle cache, we bail out. We must not proceed to acquire a connection dhandle read
+             * lock or a schema lock to acquire the HS dhandle while holding the pass lock, as this
+             * could lead to a deadlock. There are several places in the code where a pass lock is
+             * taken after a schema lock, which makes this sequence unsafe.
+             */
+            WT_RET_NOTFOUND_OK(ret = __wt_curhs_get_cached(session, hs_id, &hs_tree));
+            if (ret == 0) {
                 total_inmem += __wt_atomic_load64(&hs_tree->bytes_inmem);
                 total_dirty += __wt_atomic_load64(&hs_tree->bytes_dirty_intl) +
                   __wt_atomic_load64(&hs_tree->bytes_dirty_leaf);
                 total_updates += __wt_atomic_load64(&hs_tree->bytes_updates);
+            } else if (ret == WT_NOTFOUND) {
+                if (hs_id == WT_HS_ID)
```

Comment:
Is there benefit to seperate `cache_eviction_hs_not_cached_in_cursor` and `cache_eviction_hs_shared_not_cached_in_cursor`? Could we merge them into one?